File: class/flight_analyzer.m
%==========================================================================
% FLIGHT ANALYZER - Betaflight Controller Analysis ANALYSIS CLASS
%==========================================================================
% Purpose: 
%   Get Spectra and Spectogram out of Data
%
% Authors: 
%   Yuri Bianchi
%   Janick Dort
%   Dario Jurietti
%
% Supervisors: 
%   Michael Peter
%   Prof. Dr. Ruprecht Altenburger
%
% Date: 05.06.2026
%==========================================================================

classdef flight_analyzer
    properties       
        % Raw flight data inputs

        time             % Time vector [s]
        data             % Flgiht data
        ind              % Columns
        Ts_log           % Logging Time [s]

        % Used in different functions
        freq_spectra
        spectra
        freq_spectogram
        spectrogram_unf
        spectrogram_fil
        throttle_all

        % Default values
        linewidth            (1,1) double  = 1.2
        
        
    end

    methods
        function obj = flight_analyzer(data, ind, Ts_log)
            obj.data = data;
            obj.ind = ind;
            obj.Ts_log = Ts_log;

        end

        function obj = calculate_spectra(obj, Nestfaspec, koverlapspec)
            % =============================================================
            %  Spectral Analysis
            % =============================================================
            % Performs frequency domain analysis of gyro signals and PID sums
            % 
            % Steps:
            % 1. Configure spectral estimation parameters
            % 2. Calculate power spectra using Hann window
            % 3. Convert power to amplitude spectra
            % 4. Store results for unfiltered, filtered gyro and axis sums
                
            
            % Select data columns for spectral analysis
            data_for_spectra = obj.data(:,[obj.ind.gyroUnfilt, ...
                                       obj.ind.gyroADC, ...
                                       obj.ind.axisSum, ...
                                       obj.ind.setpoint(1:3)]);
            
            % Configure spectral estimation parameters
            Nestspec = round(Nestfaspec / obj.Ts_log);    % Window length in samples
            windowspec   = hann(Nestspec, 'periodic');        % Hann window for better frequency resolution
            Noverlap = floor(koverlapspec * Nestspec);    % Overlap for smoother estimates

            % Calculate power spectral density
            [pxx, obj.freq_spectra] = estimate_spectra(data_for_spectra, windowspec, Noverlap, Nestspec, obj.Ts_log);
            obj.spectra = sqrt(pxx); % Convert power to amplitude spectra (dc needs to be scaled differently)


        end
        function obj = calculate_spectogram(obj, Nestfaspec, koverlapspec)

            Nestspec = round(Nestfaspec / obj.Ts_log);    % Window length in samples
            windowspec   = hann(Nestspec, 'periodic');        % Hann window for better frequency resolution
            Noverlap = floor(koverlapspec * Nestspec);    % Overlap for smoother estimates

            % Calculate throttle resolution
            Nres = floor(max(obj.data(:,obj.ind.setpoint(4))) / 1e1 / 2); % should give 40 at 80% throttle constrain

            % Initialize storage for multiple spectrograms
            num_spectrograms = 3;    % One per axis
            obj.spectrogram_unf = cell(1, num_spectrograms);
            obj.spectrogram_fil = cell(1, num_spectrograms);
            obj.freq_spectogram = cell(1, num_spectrograms);
            obj.throttle_all = cell(1, num_spectrograms);

            % Calculate spectrograms for unfiltered gyro data
            for spectrogram_nr = 1:num_spectrograms
                [pxx, freq, throttle] = estimate_spectrogram( ...
                    obj.data(:, obj.ind.gyroUnfilt(spectrogram_nr)), ...    % Raw gyro data
                    obj.data(:, obj.ind.setpoint(4)) / 10.0, ...            % Throttle values
                    windowspec, Noverlap, Nestspec, Nres, obj.Ts_log);

                obj.spectrogram_unf{spectrogram_nr} = sqrt(pxx); % Convert power to amplitude
                obj.freq_spectogram{spectrogram_nr} = freq;
                obj.throttle_all{spectrogram_nr} = throttle;
            end

                 % Calculate spectrograms for filtered gyro data
            for spectrogram_nr = 1:num_spectrograms
                [pxx, freq, throttle] = estimate_spectrogram( ...
                    obj.data(:, obj.ind.gyroADC(spectrogram_nr)), ...    % Filtered gyro data
                    obj.data(:, obj.ind.setpoint(4)) / 10.0, ...         % Throttle values
                    windowspec, Noverlap, Nestspec, Nres, obj.Ts_log);

                obj.spectrogram_fil{spectrogram_nr} = sqrt(pxx); % Convert power to amplitude
                obj.freq_spectogram{spectrogram_nr} = freq;
                obj.throttle_all{spectrogram_nr} = throttle;
            end
            
        end
       
    end

end

File: py_lib/flight_analyzer.py
"""
==========================================================================
FLIGHT ANALYZER - Betaflight Controller Analysis ANALYSIS CLASS
==========================================================================

Purpose: 
    Get Spectra and Spectrogram out of Data

Authors:
    Yuri Bianchi
    Janick Dort
    Dario Jurietti

Supervisors:
    Michael Peter
    Prof. Dr. Ruprecht Altenburger

Date:
    05.06.2026
==========================================================================
"""

import numpy as np
from typing import Optional, List
from scipy.signal import hann

from .pidtuninglib import estimate_spectra, estimate_spectrogram


class FlightAnalyzer:
    """Flight data spectral analyzer."""
    
    def __init__(self, data: np.ndarray, ind, Ts_log: float):
        """
        Initialize flight analyzer.
        
        Parameters
        ----------
        data : np.ndarray
            Flight data array
        ind : ColumnIndices
            Column indices object
        Ts_log : float
            Logging sample time in seconds
        """
        self.data = data
        self.ind = ind
        self.Ts_log = Ts_log
        self.linewidth = 1.2
        
        # Storage for results
        self.freq_spectra: Optional[np.ndarray] = None
        self.spectra: Optional[np.ndarray] = None
        self.freq_spectrogram: List[np.ndarray] = []
        self.spectrogram_unf: List[np.ndarray] = []
        self.spectrogram_fil: List[np.ndarray] = []
        self.throttle_all: List[np.ndarray] = []
    
    def calculate_spectra(
        self,
        Nestfaspec: float,
        koverlapspec: float
    ) -> 'FlightAnalyzer':
        """
        Calculate power spectra of gyro and control signals.
        
        Parameters
        ----------
        Nestfaspec : float
            Window length in seconds
        koverlapspec : float
            Overlap factor (0-1)
        
        Returns
        -------
        FlightAnalyzer
            Self with calculated spectra
        """
        # Select data columns for spectral analysis
        data_for_spectra = self.data[:, np.concatenate([
            self.ind.gyroUnfilt,
            self.ind.gyroADC,
            self.ind.axisSum,
            self.ind.setpoint[:3]
        ])]
        
        # Configure spectral estimation parameters
        Nest_spec = int(np.round(Nestfaspec / self.Ts_log))
        window_spec = hann(Nest_spec, sym=False)  # periodic
        Noverlap = int(np.floor(koverlapspec * Nest_spec))
        
        # Calculate power spectral density
        pxx, self.freq_spectra = estimate_spectra(
            data_for_spectra,
            window_spec,
            Noverlap,
            Nest_spec,
            self.Ts_log
        )
        
        # Convert power to amplitude spectra
        self.spectra = np.sqrt(pxx)
        
        return self
    
    def calculate_spectrogram(
        self,
        Nestfaspec: float,
        koverlapspec: float
    ) -> 'FlightAnalyzer':
        """
        Calculate spectrograms vs throttle.
        
        Parameters
        ----------
        Nestfaspec : float
            Window length in seconds
        koverlapspec : float
            Overlap factor (0-1)
        
        Returns
        -------
        FlightAnalyzer
            Self with calculated spectrograms
        """
        Nest_spec = int(np.round(Nestfaspec / self.Ts_log))
        window_spec = hann(Nest_spec, sym=False)
        Noverlap = int(np.floor(koverlapspec * Nest_spec))
        
        # Calculate throttle resolution
        Nres = int(np.floor(np.max(self.data[:, self.ind.setpoint[3]]) / 1e1 / 2))
        
        # Initialize storage
        num_spectrograms = 3  # One per axis
        self.spectrogram_unf = []
        self.spectrogram_fil = []
        self.freq_spectrogram = []
        self.throttle_all = []
        
        # Calculate spectrograms for unfiltered gyro
        for spectrogram_nr in range(num_spectrograms):
            pxx, freq, throttle = estimate_spectrogram(
                self.data[:, self.ind.gyroUnfilt[spectrogram_nr]],
                self.data[:, self.ind.setpoint[3]] / 10.0,
                window_spec,
                Noverlap,
                Nest_spec,
                Nres,
                self.Ts_log
            )
            
            self.spectrogram_unf.append(np.sqrt(pxx))
            self.freq_spectrogram.append(freq)
            self.throttle_all.append(throttle)
        
        # Calculate spectrograms for filtered gyro
        for spectrogram_nr in range(num_spectrograms):
            pxx, freq, throttle = estimate_spectrogram(
                self.data[:, self.ind.gyroADC[spectrogram_nr]],
                self.data[:, self.ind.setpoint[3]] / 10.0,
                window_spec,
                Noverlap,
                Nest_spec,
                Nres,
                self.Ts_log
            )
            
            self.spectrogram_fil.append(np.sqrt(pxx))
        
        return self